src/librustc/traits/select.rs

```diff
@@ -1710,22 +1710,44 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
     {
         debug!("assemble_candidates_from_impls(obligation={:?})", obligation);
 
+        // Check if default impls should be emitted.
+        // default impls are emitted if the param_env is refered to a default impl.
+        // The param_env should contain a Self: Trait<..> predicate in those cases
+        let self_trait_is_present:Vec<&ty::Predicate<'tcx>> =
```

Comment:
Hmm, I don't think we should have to modify select.rs at all. Instead, what I would expect is for us to modify the set of predicates associated with default impls, analogously to what we do with traits. This would be part of the `predicates_of` query -- for example, in traits, we insert `Self: Trait<...>` right here:

https://github.com/rust-lang/rust/blob/4d2d3fc5dadf894a8ad709a5860a549f2c0b1032/src/librustc_typeck/collect.rs#L1444-L1446

I would think that we would want to do the same code in the case of a `default impl`. In that case, I don't think we should have to modify `select.rs` at all. The required assumptions will be found in the environment.